electrum/submarine_swaps.py

```diff
@@ -541,3 +534,50 @@ def get_send_amount(self, recv_amount: Optional[int], *, is_reverse: bool) -> Op
         if is_reverse and send_amount is not None:
             send_amount += self.get_claim_fee()
         return send_amount
+
+    def get_swap_by_tx(self, tx: Transaction) -> Optional[SwapData]:
+        # determine if tx is spending from a swap
+        txin = tx.inputs()[0]
+        return self.get_swap_by_claim_txin(txin)
+
+    def get_swap_by_claim_txin(self, txin: TxInput) -> Optional[SwapData]:
+        for key, swap in self.swaps.items():
+            if txin.prevout.txid.hex() == swap.funding_txid:
+                return swap
+        return None
```

Comment:
It is insufficient to compare just the txid to identify claim txs. The full outpoint needs to be compared (txid+out_idx).

Consider doing two forward swaps in a row.
tx1 funds swap1 and also creates a change output.
Now we start doing swap2, and the previous change output is chosen as an input for funding tx2.
tx2 which is funding swap2 is going to be incorrectly identified as a claim tx for swap1, as
`tx2.intputs()[0].prevout.txid.hex() == swap1.funding_txid`
but note that
`tx2.intputs()[0].prevout.out_idx != swap1.funding_outidx` (or something... the swap object is missing the required field)

Reply:
Indeed, this is relevant for refund transactions. 

Note that since 1364e7538acf37d1b649d2ca8ef647858540b227, the `funding_txid` field of `SwapData` is populated by `lnwatcher`, and does not need to be persisted anymore. Without a wallet_db upgrade, we could populate that field with the serialized outpoint, or add a new, non-persisted field to `SwapData`, that would have the proper name.

(Not sure if that would still work with  `subscribe_to_outpoint`)

Reply:
There is a pending db upgrade for `SwapData` in the `seamless_lightning` branch, that persists `receive_address`.
Maybe we should move that part of the db upgrade out of that branch, and bundle it whatever we need here.

Reply:
This would also work, I guess (without db upgrade):
```python
--- a/electrum/submarine_swaps.py
+++ b/electrum/submarine_swaps.py
@@ -92,6 +92,7 @@ class SwapData(StoredObject):
     funding_txid = attr.ib(type=Optional[str])
     spending_txid = attr.ib(type=Optional[str])
     is_redeemed = attr.ib(type=bool)
+    _funding_prevout = None
 
 
 def create_claim_tx(
@@ -171,6 +172,7 @@ class SwapManager(Logger):
                 self.logger.info('amount too low, we should not reveal the preimage')
                 continue
             swap.funding_txid = txin.prevout.txid.hex()
+            swap._funding_prevout = txin.prevout
             spent_height = txin.spent_height
             if spent_height is not None:
                 swap.spending_txid = txin.spent_txid
@@ -542,7 +544,7 @@ class SwapManager(Logger):
 
     def get_swap_by_claim_txin(self, txin: TxInput) -> Optional[SwapData]:
         for key, swap in self.swaps.items():
-            if txin.prevout.txid.hex() == swap.funding_txid:
+            if txin.prevout == swap._funding_prevout:
                 return swap
         return None
```